xprof/interval.c.erb

```diff
@@ -235,24 +234,31 @@ bt_message_iterator_class_next_method_status <%= namespace %>_dispatch_message_i
 
 consume_upstream_messages:
 
-    next_status = bt_message_iterator_next(dispatch_iter->message_iterator,
-        &upstream_messages, &upstream_message_count);
+    next_status = BT_MESSAGE_ITERATOR_NEXT_STATUS_OK;
+
+    if (!<%= namespace %>_last_call) {
+      // We cannot call bt_message_iterator_next if previously returned BT_MESSAGE_ITERATOR_NEXT_STATUS_END
+      next_status = bt_message_iterator_next(dispatch_iter->message_iterator,
+            &upstream_messages, &upstream_message_count);
+      if (next_status == BT_MESSAGE_ITERATOR_NEXT_STATUS_END) {
+          <%= namespace %>_last_call = true;
+          bt_message *message = bt_message_stream_end_create(self_message_iterator, dispatch_iter->dispatch->stream);
+          <%= namespace %>_downstream_message_queue_push(dispatch_iter, message);
+      }
+    }
 
     /* Initialize the return status to a success */
     bt_message_iterator_class_next_method_status status =
-        BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_OK;
-
-    if (next_status == BT_MESSAGE_ITERATOR_NEXT_STATUS_END &&  !<%= namespace %>_last_call) {
-      bt_message *message = bt_message_stream_end_create(self_message_iterator, dispatch_iter->dispatch->stream);
-      <%= namespace %>_downstream_message_queue_push(dispatch_iter, message);
-      <%= namespace %>_last_call = true;
-    }
+      BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_OK;
 
-    if ( (next_status == BT_MESSAGE_ITERATOR_NEXT_STATUS_END) && <%= namespace %>_downstream_message_queue_empty(dispatch_iter) ) {
+    // No more message to send, we can exit
+    if (<%= namespace %>_last_call && <%= namespace %>_downstream_message_queue_empty(dispatch_iter) )  {
        bt_message_iterator_put_ref(dispatch_iter->message_iterator);
        status = BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_END;
        goto end;
     }
+    // We really on the fact that if <%= namespace %>_last_call> but we still have somne message to send,
+    // bt2 will put upstream_message_count == 0, hence we will bypass the reading step
```

Comment:
AFAIK, when we call: 

`next_status = bt_message_iterator_next(dispatch_iter->message_iterator, &upstream_messages, &upstream_message_count);`.

* `next_status` could be `BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_END`.
* We still can have messages in `upstream_messages` and `upstream_message_count > 0`.

So we can not put the stream end message in the queue, since we still have messages in `upstream_messages`, **so we need to put the remaining upstream messages, then the end message?**

```c
bt_message *message = bt_message_stream_end_create(self_message_iterator, dispatch_iter->dispatch->stream);
<%= namespace %>_downstream_message_queue_push(dispatch_iter, message);
```

Reply:
I don't think  BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_END can return messages, and messages are only returned when BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_OK. But I think this is the only thing we're certain about.

Reply:
Yeah, we should. But don't want to invest to much time on this as it's already  cleaned and better on metababel.

Reply:
No, I'm stupid. We are setting it to `0` before. We are good.